src/constructLayoutEngine.js

```diff
@@ -0,0 +1,164 @@
+ * isActive() => boolean;
+ * }} LayoutEngine
+ *
+ * @typedef {{
+ * routes: import('./constructRoutes').ResolvedRoutesConfig;
+ * applications: Array<import('single-spa').RegisterApplicationConfig & import('./constructApplications').WithLoadFunction>;
+ * }} LayoutEngineOptions
+ *
+ * @param {LayoutEngineOptions} layoutEngineOptions
+ * @returns {LayoutEngine}
+ */
+export function constructLayoutEngine({
+  routes: resolvedRoutes,
+  applications,
+}) {
+  let active = false;
+
+  const layoutEngine = {
+    isActive: () => active,
+    activate() {
+      if (active) {
+        return;
+      } else {
+        active = true;
+      }
+
+      if (inBrowser) {
+        window.addEventListener(
+          "single-spa:before-mount-routing-event",
+          arrangeDomElements
+        );
+
+        arrangeDomElements();
+      }
+    },
+    deactivate() {
+      if (!active) {
+        return;
+      } else {
+        active = false;
+      }
+
+      if (inBrowser) {
+        window.removeEventListener(
+          "single-spa:before-mount-routing-event",
+          arrangeDomElements
+        );
+      }
+    },
+  };
+
+  layoutEngine.activate();
+
+  return layoutEngine;
+
+  function arrangeDomElements() {
+    const path = location[resolvedRoutes.mode === "hash" ? "hash" : "pathname"];
+
+    // TODO memoize this for perf
+    const matchedRoutes = matchRoute(resolvedRoutes, path);
+
+    const initialContainer =
+      typeof resolvedRoutes.containerEl === "string"
+        ? document.querySelector(resolvedRoutes.containerEl)
+        : resolvedRoutes.containerEl;
+
+    recurseRoutes(matchedRoutes.routes, initialContainer);
+  }
+}
+
+/**
+ *
+ * @param {Array<import('./constructRoutes').Route>} routes
+ * @param {HTMLElement} parentContainer
+ * @param {HTMLElement=} previousSibling
+ * @returns {HTMLElement}
+ */
+function recurseRoutes(routes, parentContainer, previousSibling) {
+  routes.forEach((route) => {
+    if (route.type === "application") {
+      const applicationContainer = getContainerEl(parentContainer, route);
+      const applicationElement = getApplicationElement(route);
+
+      if (
+        previousSibling &&
+        previousSibling.parentNode === applicationContainer
+      ) {
+        // move to be immediately after previousSibling
+        previousSibling.insertAdjacentElement("afterend", applicationElement);
+      } else if (applicationElement.parentNode !== applicationContainer) {
+        // append to end of the container
+        applicationContainer.appendChild(applicationElement);
+      }
+
+      // Only use this as the reference sibling node if it's within the parent container
+      if (applicationContainer === parentContainer) {
+        previousSibling = applicationElement;
+      }
+    } else {
+      previousSibling = recurseRoutes(
+        route.routes,
+        getContainerEl(parentContainer, route),
+        previousSibling
+      );
+    }
+
+    return previousSibling;
+  });
+}
```

Comment:
Again it's just personal preference but I like having tests for methods that use recursion directly (even if they are tested by the public API), but it's a preference so don't feel obligated to add them.

My preference is because recursion is complicated and it usually takes me some time to grok recursion (even recursion that I've written) and I like having the safety of additional unit tests

Reply:
my preference is to test the public API